File: 006LinerList/Java/src/MyList.java
import java.util.Optional;
import java.util.function.Function;

/**
 * Created by naoki on 2015/07/09.
 */
public interface MyList<T>{

    public Optional<T> get(int index);

    public int size();

    public default MyList<T> add(T value){
        return new Cons<T>(value,this);
    }

    public default MyList<T> reverse(){
       MyList<T> list = new Nil<T>();
        int size = size();
        for(int i = 0;i <size;i++){

            list = get(i).equals(Optional.<T>empty())?list:list.add(get(i).get());

        }
        return list;
    }

    public default MyList<T> apend(MyList<T> list){
        int index = list.size()-1;
        MyList<T> rList = this;
        while(index >= 0){
            rList = rList.add(list.get(index).get());
            index--;
        }
        return rList;
    }

    public default <R> MyList<R> map(Function<T,R> function){

        MyList<R> rList = new Nil<R>();
        for(int i = this.size()-1;i>=0;i--){
            rList = rList.add(function.apply(this.get(i).get()));
        }
        return rList;
    }

    public default MyList<T> take(int index){
        MyList<T> rList = new Nil<T>();

        if(index > size()-1||index < 0)throw new IllegalArgumentException();

        for(int i = index;i>0;i--){

            rList = rList.add(this.get(i-1).get());
        }
        return rList;
    }

    public default MyList<T> drop(int index){
        MyList<T> rList = new Nil<T>();
        int size = size();

        if(index > size-1||index<0)throw new IllegalArgumentException();

        for(int i = size-1;i>index-1;i--){
            rList = rList.add(this.get(i).get());
        }
        return rList;
    }
}

File: 006LinerList/Scala/src/main/scala/MyList.scala
/**
 * Created by naoki on 2015/07/08.
 */
sealed trait MyList[+A] {

  def get(index:Int):Option[A] =
    this match{
      case Nil => None
      case Cons(head,tail) =>
        if(index <= 0)
          Some(head)
        else
          tail.get(index -1 )
  }

  def add[B>:A](value:B):MyList[B] ={
    new Cons[B](value,this)
  }

  def size:Int =
    this match{
      case Cons(_,tail) => 1 + tail.size
      case Nil => 0
    }

  def reverse[B>:A]:MyList[B]={
    this match{
      case Cons(head,tail) => tail.reverse(Nil.add(head))
      case Nil => Nil
    }
  }
  private def reverse[B>:A](list:MyList[B]):MyList[B]=this match{
    case Cons(head,tail) => tail.reverse(list.add(head))
    case Nil => list
  }


  def append[B>:A](list:MyList[B]):MyList[B]=list match{
    case Cons(head,tail)=> apend(tail).add(head)
    case Nil => this
  }

  def map[B](f:(A=>B)):MyList[B]=this match{
    case Cons(head,tail)=> tail.map(f).add(f(head))
    case Nil => Nil
  }

  def drop(index:Int):MyList[A]=this match{
    case Cons(head,tail)=>{
      if(index<=0)tail.drop(index).add(head)
      else tail.drop(index-1)
    }
    case Nil => Nil
  }

  def take(index:Int):MyList[A]=this match{
    case Cons(head,tail)=>{
      if(index<=0)Nil
      else tail.take(index-1).add(head)
    }
    case Nil => Nil
  }


}

object MyList{
  def apply[A](values:A*):MyList[A] =
    values.foldRight(Nil:MyList[A])(Cons(_,_))
}

case class Cons[A](head:A,tail:MyList[A]) extends MyList[A]

case object Nil extends MyList[Nothing]


